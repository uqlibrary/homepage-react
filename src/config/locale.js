const CAMPUS_ST_LUCIA = 'St Lucia';
const CAMPUS_GATTON = 'Gatton';
const CAMPUS_HERSTON = 'Herston';
const CAMPUS_DUTTON_PARK = 'Dutton Park';

export const locale = {
    locations: [
        {
            displayName: 'No preference',
            value: 'not set',
        },
        {
            displayName: 'St Lucia',
            value: CAMPUS_ST_LUCIA,
        },
        {
            displayName: 'Gatton',
            value: CAMPUS_GATTON,
        },
        {
            displayName: 'Herston',
            value: CAMPUS_HERSTON,
        },
        {
            displayName: 'Dutton Park',
            value: CAMPUS_DUTTON_PARK,
        },
    ],
    hoursCampusMap: {
        // the keys in this must match the abbr values found in the Locations api result
        'Arch Music': CAMPUS_ST_LUCIA,
        AskUs: 'Online',
        'Biol Sci': CAMPUS_ST_LUCIA,
        Central: CAMPUS_ST_LUCIA,
        DHEngSci: CAMPUS_ST_LUCIA,
        'Duhig Study': CAMPUS_ST_LUCIA,
        Fryer: CAMPUS_ST_LUCIA,
        Gatton: CAMPUS_GATTON,
        Herston: CAMPUS_HERSTON,
        Law: CAMPUS_ST_LUCIA,
        'Whitty Mater': 'Other',
        'Dutton Park': CAMPUS_DUTTON_PARK,
        Bundaberg: 'Other',
        HerveyBay: 'Other',
        Rockhampton: 'Other',
        Toowoomba: 'Other',
    },

    // this table maps those locations who exist on vemcount against their matching springshare location
    // note: not all locations have vemcount people-counting gates
    vemcountSpringshareMapping: [
        {
            springshareId: 3967,
            vemcountZoneId: 7880,
            name: 'Dutton park', // this doesn't need to match either system, its for the developer to not have to track raw numbers
        },
        {
            springshareId: 3842,
            vemcountZoneId: 7665,
            name: 'Central',
        },
        {
            springshareId: 3823,
            vemcountZoneId: 7877,
            name: 'Architecture',
        },
        {
            springshareId: 3824,
            vemcountZoneId: 7878,
            name: 'BSL',
        },
        {
            springshareId: 3825,
            vemcountZoneId: 7879,
            name: 'DHESL',
        },
        {
            springshareId: 3830,
            vemcountZoneId: 7411,
            name: 'Duhig tower',
        },
        {
            springshareId: 3833,
            vemcountZoneId: 7884,
            name: 'Gatton',
        },
        {
            springshareId: 3838,
            vemcountZoneId: 7883,
            name: 'Herston',
        },
        {
            springshareId: 3841,
            vemcountZoneId: 7882,
            name: 'Law',
        },
    ],
};
