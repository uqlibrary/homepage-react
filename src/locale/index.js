import { default as validationErrors } from './validationErrors';
import { default as forms } from './forms';
import { default as pages } from './pages';
import { default as global } from './global';
import { default as menu } from './menu';
import { default as templates } from './templates';

export const locale = {
    ...validationErrors,
    ...forms,
    ...pages,
    ...global,
    ...menu,
    ...templates,
};
