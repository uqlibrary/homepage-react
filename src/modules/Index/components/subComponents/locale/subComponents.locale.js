export default {
    referencingPanel: {
        items: [
            {
                title: 'Referencing style guides',
                url: 'https://guides.library.uq.edu.au/referencing',
                description: 'APA, Chicago, Vancouver and more',
                uqOnly: false,
            },
            {
                title: 'Endnote referencing software',
                url: 'https://web.library.uq.edu.au/research-tools-techniques/endnote-referencing-software',
                description: 'Download and support',
                uqOnly: true,
            },
        ],
    },
};
