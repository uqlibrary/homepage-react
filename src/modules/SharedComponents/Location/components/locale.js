export const locale = {
    tooltip: 'Current preferred location is [currentLocation] - Click to change',
    noLocationSet: 'not set',
    noLocationSetLabel: 'Set a preferred campus',
    menuTitle: 'Select a preferred location',
    locations: [
        {
            location: 'No preference',
            value: 'not set',
        },
        {
            location: 'St Lucia',
            value: 'St Lucia',
        },
        {
            location: 'Gatton',
            value: 'Gatton',
        },
        {
            location: 'Herston',
            value: 'Herston',
        },
    ],
};
