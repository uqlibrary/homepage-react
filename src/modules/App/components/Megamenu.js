import React, { createRef, useEffect } from 'react';
import PropTypes from 'prop-types';

import { default as menuLocale } from 'locale/menu';
import classNames from 'classnames';

import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Collapse from '@material-ui/core/Collapse';
import ExpandLess from '@material-ui/icons/ExpandLess';
import ExpandMore from '@material-ui/icons/ExpandMore';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => {
    return {
        megamenublock: {
            backgroundColor: theme.palette.white.main,
            width: '100%',
        },
        mainMenu: {
            outline: 'none',
            paddingTop: 0,
            [theme.breakpoints.up('lg')]: {
                display: 'flex',
                margin: 0,
                paddingTop: 0,
                paddingBottom: 0,
                paddingLeft: 16,
                '& > span > div': {
                    paddingRight: 12,
                    paddingLeft: 0,
                },
                '& > span > div > div': {
                    paddingLeft: 6,
                },
                // remove the flash to grey when we mouse over a top menu
                '& span > div[role="button"]:hover': {
                    backgroundColor: 'initial',
                },
                '& > div:first-child': {
                    marginLeft: 28,
                },
                '& > div > div': {
                    paddingLeft: 8,
                    paddingRight: 8,
                },
                // first menu item should be lined up with Library title
                '& > div:first-child > div': {
                    paddingLeft: 0,
                },
            },
        },
        ListItemTextPrimary: {
            whiteSpace: 'nowrap',
            fontWeight: 400,
        },
        ListItemTextSecondary: {
            ...theme.typography.caption,
        },
        menuDropdown: {
            backgroundColor: theme.palette.secondary.light,
            [theme.breakpoints.up('lg')]: {
                // apply this to mobile as well and the submenu goes over the other menu headers
                zIndex: 1000,
                position: 'absolute',
            },
        },
        shiftLeft: {
            marginLeft: '-20rem',
        },
        submenus: {
            [theme.breakpoints.up('lg')]: {
                flexDirection: 'row',
            },
        },
        menuColumns: {
            [theme.breakpoints.up('lg')]: {
                display: 'flex',
            },
            '& > div': {
                marginTop: '6px',
                borderLeft: '1px solid #ccc',
            },
            '& div:first-child': {
                borderLeft: 'none',
            },
        },
        currentPage: {
            borderBottom: '2px solid #51247a',
        },
        menuItemContainer: {
            [theme.breakpoints.up('lg')]: {
                '& > div': {
                    paddingLeft: 0,
                },
            },
            [theme.breakpoints.down('md')]: {
                '& div': {
                    margin: 0,
                    padding: 0,
                },
                '& > div > div:first-child > span': {
                    // top level of menu under hamburger
                    border: '1px solid #e2e2e2',
                    padding: '1rem 1.5rem',
                },
                '& div > div:first-child div > span': {
                    // primaryText of submenu items
                    borderBottom: '1px solid #e2e2e2',
                    padding: '0.7rem 1.5rem 0.7rem 2.5rem',
                },
                '& svg': {
                    border: '1px solid #e2e2e2',
                    padding: '1rem',
                },
            },
        },
        menuItem: {
            [theme.breakpoints.up('lg')]: {
                paddingTop: 0,
                paddingBottom: 0,
                '& div': {
                    marginTop: 0,
                    minHeight: 45,
                },
            },
        },
    };
};

export function Megamenu(props) {
    const { classes, docked, menuOpen, menuItems, toggleMenu, history, isMobile, ...rest } = props;

    // from https://usehooks.com/useOnClickOutside/
    function useOnClickOutside(ref, handler) {
        useEffect(
            () => {
                const listener = event => {
                    // Do nothing if clicking ref's element or descendent elements
                    if (!ref.current || ref.current.contains(event.target)) {
                        return;
                    }

                    handler(event);
                };

                document.addEventListener('mousedown', listener);
                document.addEventListener('touchstart', listener);

                return () => {
                    document.removeEventListener('mousedown', listener);
                    document.removeEventListener('touchstart', listener);
                };
            },
            // Add ref and handler to effect dependencies
            // It's worth noting that because passed in handler is a new ...
            // ... function on every render that will cause this effect ...
            // ... callback/cleanup to run every render. It's not a big deal ...
            // ... but to optimize you can wrap handler in useCallback before ...
            // ... passing it into this hook.
            [ref, handler],
        );
    }

    const menuRef = createRef();

    // an array so we can control each submenu separately
    const initialSubMenus = [];
    const defaultMenuPosition = false;
    menuItems.forEach(item => (initialSubMenus[item.id] = defaultMenuPosition));
    const [isSubMenuOpen, setSubMenuOpen] = React.useState(initialSubMenus);

    function setBackGroundColourOfMenuHeader(id, color) {
        document.getElementById(id).style.backgroundColor = color;
    }

    const setParticularSubMenuOpen = (changingId, newOpen) => {
        // using an array allows us to update one element of the open/closed array, to match one menuitem change
        const newValues = [];
        Object.keys(isSubMenuOpen).forEach(key => {
            newValues[key] = key === changingId ? newOpen : initialSubMenus[key];

            // set the menu label to match background colour
            setBackGroundColourOfMenuHeader(key, newOpen && key === changingId ? '#f2f2f2' : '#fff');
        });
        setSubMenuOpen(newValues);

        // scroll up when they change menus, otherwise we often end with the menu off screen :(
        if (isMobile) {
            const topHeight = document.getElementById('uqheader').offsetHeight;
            document.getElementById('content-container').scrollTop = topHeight;
        }
    };

    const isAnyMenuOpen = () => {
        return !!Object.values(isSubMenuOpen).find(item => {
            return item === true;
        });
    };

    const closeAllSubMenus = () => {
        if (!isAnyMenuOpen()) {
            return false;
        }
        const newValues = [];
        Object.keys(isSubMenuOpen).forEach(key => {
            newValues[key] = defaultMenuPosition;
            // set the menu header back to default colour
            setBackGroundColourOfMenuHeader(key, '#fff');
        });
        setSubMenuOpen(newValues);
        return true;
    };

    useOnClickOutside(menuRef, () => closeAllSubMenus(false));

    // from https://stackoverflow.com/questions/37440408/how-to-detect-esc-key-press-in-react-and-how-to-handle-it/46123962
    const catchKeyClick = event => {
        // close menu when escape key pressed
        const escapeKeyCode = 27;
        if (event.keyCode === escapeKeyCode) {
            closeAllSubMenus(false);
        }

        const tabKeyCode = 9;
        // if they tab out the end of a submenu, close the menu

        /* istanbul ignore next */
        if (event.keyCode === tabKeyCode) {
            const focusedItem = document.activeElement;
            if (
                !!document.getElementById(focusedItem.id) &&
                !!document.getElementById(focusedItem.id).classList &&
                document.getElementById(focusedItem.id).classList.contains('endmenuItem')
            ) {
                closeAllSubMenus(false);
            }
        }

        // if they shift-tab back out of a menu, close the menu
        /* istanbul ignore next */
        if (event.shiftKey && event.keyCode === tabKeyCode) {
            const focusedItem = document.activeElement;
            if (
                !!document.getElementById(focusedItem.id) &&
                !!document.getElementById(focusedItem.id).classList &&
                document.getElementById(focusedItem.id).classList.contains('submenuheader')
            ) {
                closeAllSubMenus(false);
            }
        }
    };
    useEffect(() => {
        document.addEventListener('keydown', catchKeyClick, false);

        return () => {
            document.removeEventListener('keydown', catchKeyClick, false);
        };
    });

    const focusOnElementId = elementId => {
        if (document.getElementById(elementId)) {
            document.getElementById(elementId).focus();
        }
    };

    const navigateToLink = (url, target = null) => {
        const isInternaLink = url => url.indexOf('http') === -1;
        if (!!url) {
            if (isInternaLink(url)) {
                history.push(url);
            } else if (target !== null) {
                // external link
                window.open(url, target);
            } else {
                window.location.assign(url);
            }
        }
    };

    function clickMenuItem(menuItem) {
        return !!menuItem.submenuItems && menuItem.submenuItems.length > 0
            ? setParticularSubMenuOpen(menuItem.id, !isSubMenuOpen[menuItem.id])
            : navigateToLink(menuItem.linkTo, menuItem.target || null);
    }

    function renderSingleColumn(index, classes, menuColumn, isLastColumn) {
        return (
            <List
                component="div"
                disablePadding
                key={`megamenu-group-${index}`}
                data-testid={`megamenu-group-${index}`}
                id={`megamenu-group-${index}`}
                className={classes.verticalMenuList}
            >
                {!!menuColumn &&
                    menuColumn.length > 0 &&
                    menuColumn.map((submenuItem, index2) => {
                        const endItemClass = isLastColumn && index2 === menuColumn.length - 1 ? 'endmenuItem' : '';
                        return (
                            <ListItem
                                button
                                data-testid={`megamenu-group-${index}-item-${index2}`}
                                key={`megamenu-group-${index}-item-${index2}`}
                                id={`megamenu-group-${index}-item-${index2}`}
                                onClick={() => navigateToLink(submenuItem.linkTo, submenuItem.target || null)}
                                className={classNames(classes.menuItem, `${endItemClass}`)}
                            >
                                <ListItemText
                                    classes={{
                                        primary: classes.ListItemTextPrimary,
                                        secondary: classes.ListItemTextSecondary,
                                    }}
                                    primary={submenuItem.primaryText}
                                    secondary={!isMobile && submenuItem.secondaryText}
                                />
                            </ListItem>
                        );
                    })}
            </List>
        );
    }

    const renderSubMenu = (menuItem, index, classes) => {
        const menuColumns = [];
        menuItem.submenuItems.map(submenuItem => {
            const index = submenuItem.column || 1;
            if (!menuColumns[index]) {
                menuColumns[index] = [];
            }
            menuColumns[index].push(submenuItem);
        });

        return (
            <Collapse
                in={isSubMenuOpen[menuItem.id]}
                timeout="auto"
                unmountOnExit
                className={classNames(!!menuItem.shiftLeft ? classes.shiftLeft : '', classes.menuDropdown)}
            >
                <div className={classes.menuColumns} data-testid={`submenu-${index}`}>
                    {menuColumns.length > 0 &&
                        menuColumns.map((menuColumn, index1) => {
                            return renderSingleColumn(index1, classes, menuColumn, index1 === menuColumns.length - 1);
                        })}
                </div>
            </Collapse>
        );
    };

    const renderSingleMenu = (menuItem, index) => {
        const hasChildren = !!menuItem.submenuItems && menuItem.submenuItems.length > 0;
        const iconSize = isMobile ? 'default' : 'small';
        return (
            <div className={classes.menuItemContainer} key={`menucontainer-item-${index}`} id={menuItem.id}>
                <ListItem
                    button
                    className={classNames(
                        menuItem.linkTo === window.location.href ? classes.currentPage : '',
                        'submenuheader',
                    )}
                    data-testid={`megamenu-submenus-item-${index}`}
                    key={`megamenu-submenus-item-${index}`}
                    id={`megamenu-submenus-item-${index}`}
                    onClick={() => clickMenuItem(menuItem)}
                >
                    <ListItemText
                        classes={{
                            primary: classes.ListItemTextPrimary,
                            secondary: classes.ListItemTextSecondary,
                        }}
                        primary={menuItem.primaryText}
                        secondary={menuItem.secondaryText}
                    />
                    {hasChildren && isSubMenuOpen[menuItem.id] && <ExpandLess size={iconSize} color="primary" />}
                    {hasChildren && !isSubMenuOpen[menuItem.id] && <ExpandMore size={iconSize} color="primary" />}
                </ListItem>
                {hasChildren && renderSubMenu(menuItem, index, classes)}
            </div>
        );
    };

    if (menuOpen && !docked) {
        // set focus on menu on mobile view if menu is opened
        setTimeout(focusOnElementId.bind(this, 'mainMenu'), 0);
    }

    if (!menuOpen) {
        return <div data-testid="mega-menu-empty" className="megamenu empty" />;
    }

    const renderHomePageItem = () => {
        return (
            <ListItem
                button
                className={classes.submenus}
                data-testid={menuLocale.menuhome.dataTestid}
                key={menuLocale.menuhome.dataTestid}
                id={menuLocale.menuhome.dataTestid}
                onClick={() => navigateToLink(menuLocale.menuhome.linkTo || null)}
            >
                <ListItemText
                    classes={{
                        primary: classes.ListItemTextPrimary,
                    }}
                    primary={menuLocale.menuhome.primaryText}
                />
            </ListItem>
        );
    };

    function renderCloseItem() {
        return (
            <ListItem
                button
                className={classes.submenus}
                data-testid={menuLocale.responsiveClose.dataTestid}
                key={menuLocale.responsiveClose.dataTestid}
                id={menuLocale.responsiveClose.dataTestid}
                onClick={() => toggleMenu}
            >
                <ListItemText
                    classes={{
                        primary: classes.ListItemTextPrimary,
                    }}
                    primary={menuLocale.responsiveClose.primaryText}
                />
            </ListItem>
        );
    }

    return (
        <div className={classes.megamenublock} data-testid="mega-menu" id={rest.id || 'megamenu'}>
            <List component="nav" data-testid="main-menu" id="mainMenu" className={classes.mainMenu} ref={menuRef}>
                {props.hasHomePageItem && renderHomePageItem()}
                {menuItems.map((menuItem, index) => {
                    return renderSingleMenu(menuItem, index);
                })}
                {props.hasCloseItem && renderCloseItem()}
            </List>
        </div>
    );
}

Megamenu.propTypes = {
    hasCloseItem: PropTypes.bool,
    hasHomePageItem: PropTypes.bool,
    logoLink: PropTypes.string,
    menuItems: PropTypes.array.isRequired,
    menuOpen: PropTypes.bool,
    isMobile: PropTypes.bool,
    docked: PropTypes.bool,
    toggleMenu: PropTypes.func,
    history: PropTypes.object.isRequired,
    classes: PropTypes.object,
};

Megamenu.defaultProps = {
    hasCloseItem: false,
    hasHomePageItem: false,
    menuOpen: true,
    isMobile: false,
};

export function isSame(prevProps, nextProps) {
    return (
        nextProps.menuOpen === prevProps.menuOpen &&
        JSON.stringify(nextProps.menuItems) === JSON.stringify(prevProps.menuItems) &&
        nextProps.docked === prevProps.docked
    );
}

export default React.memo(withStyles(styles, { withTheme: true })(Megamenu), isSame);
