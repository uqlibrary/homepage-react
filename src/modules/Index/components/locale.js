import React from 'react';
import AccessTimeIcon from '@material-ui/icons/AccessTime';
import ImportantDevicesIcon from '@material-ui/icons/ImportantDevices';
import InfoOutlined from '@material-ui/icons/InfoOutlined';
import { searchPanelLocale } from 'modules/Index/components/SearchPanel/components/searchPanelLocale';

export default {
    homePageSearch: searchPanelLocale,
    LibraryServices: {
        links: [
            {
                id: 'servicesforcommunity',
                title: 'Services for community',
                url: 'https://web.library.uq.edu.au/library-services/services-for-community',
            },
            {
                id: 'servicesforhdrs',
                title: 'Services for HDR students',
                url: 'https://web.library.uq.edu.au/library-services/services-higher-degree-by-research',
            },
            {
                id: 'servicesforhospital',
                title: 'Services for hospital staff',
                url: 'https://web.library.uq.edu.au/library-services/services-for-hospital-staff',
            },
            {
                id: 'servicesforprofessional',
                title: 'Services for professional staff',
                url: 'https://web.library.uq.edu.au/library-services/services-professional-staff',
            },
            {
                id: 'servicesforresearchers',
                title: 'Services for researchers',
                url: 'https://web.library.uq.edu.au/library-services/services-researchers',
            },
            {
                id: 'servicesforsecondary',
                title: 'Services for secondary schools',
                url: 'https://web.library.uq.edu.au/library-services/services-for-secondary-schools',
            },
            {
                id: 'servicesforstudents',
                title: 'Services for students',
                url: 'https://web.library.uq.edu.au/library-services/services-students',
            },
            {
                id: 'servicesforteaching',
                title: 'Services for teaching staff',
                url: 'https://web.library.uq.edu.au/library-services/services-teaching-staff',
            },
            {
                id: 'servicesforalumni',
                title: 'Services for UQ alumni',
                url: 'https://web.library.uq.edu.au/library-services/services-uq-alumni',
            },
            {
                id: 'servicesforexternal',
                title: 'Services for external and remote students',
                url:
                    'https://web.library.uq.edu.au/borrowing-requesting/how-borrow/borrowing-external-and-remote-students',
            },
            {
                id: 'ithelp',
                title: 'IT Help',
                url: 'https://web.library.uq.edu.au/library-services/it',
            },
            {
                id: 'digitalessentials',
                title: 'Improve your digital skills',
                url: 'https://web.library.uq.edu.au/research-tools-techniques/digital-essentials',
            },
        ],
    },
    Hours: [
        {
            title: 'Arch music',
            link: '#',
            location: '',
            hours: '7:30am-7:30pm',
            icon: null,
            iconInfo: null,
        },
        {
            title: 'AskUs',
            link: '#',
            location: '',
            hours: '8am-8pm',
            icon: null,
            iconInfo: null,
        },
        {
            title: 'Biol Sci',
            link: '#',
            location: '',
            hours: '24hrs',
            icon: <AccessTimeIcon style={{ color: 'green' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'Bundaberg',
            link: '#',
            location: '',
            hours: '8:30am-4:30am',
            icon: null,
            iconInfo: null,
        },
        {
            title: 'Central',
            link: '#',
            location: '',
            hours: '24hrs',
            icon: <AccessTimeIcon style={{ color: 'green' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'DHEngSci',
            link: '#',
            location: '',
            hours: '24hrs',
            icon: <AccessTimeIcon style={{ color: 'green' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'Duhig Study',
            link: '#',
            location: '',
            hours: '24hrs',
            icon: <AccessTimeIcon style={{ color: 'green' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'Fryer',
            link: '#',
            location: '',
            hours: '9am-5pm by appointment',
            icon: <AccessTimeIcon style={{ color: 'orange' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'Gatton',
            link: '#',
            location: '',
            hours: '24hrs',
            icon: <AccessTimeIcon style={{ color: 'green' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'Herston',
            link: '#',
            location: '',
            hours: 'Chat/Email/Phone',
            icon: <ImportantDevicesIcon style={{ color: 'orange' }} />,
            iconInfo: 'Online via Chat/Email/Phone only',
        },
        {
            title: 'Harvey Bay',
            link: '#',
            location: '',
            hours: 'Online/Phone',
            icon: <ImportantDevicesIcon style={{ color: 'orange' }} />,
            iconInfo: 'Online via Chat/Phone only',
        },
        {
            title: 'Law',
            link: '#',
            location: '',
            hours: '24hrs',
            icon: <AccessTimeIcon style={{ color: 'green' }} />,
            iconInfo: 'Part of this location is open 24hrs',
        },
        {
            title: 'PACE',
            link: '#',
            location: '',
            hours: '7am-10:30pm',
            icon: <InfoOutlined style={{ color: '#316799' }} />,
            iconInfo: 'Extended study space hours',
        },
        {
            title: 'Rockhampton',
            link: '#',
            location: '',
            hours: 'Online/Phone',
            icon: <ImportantDevicesIcon style={{ color: 'orange' }} />,
            iconInfo: 'Online via Chat/Phone only',
        },
        {
            title: 'Toowoomba',
            link: '#',
            location: '',
            hours: '8:30am-4:30pm',
            icon: null,
        },
        {
            title: 'Whitty Mater',
            link: '#',
            location: '',
            hours: '6:30am-10pm',
            icon: null,
        },
    ],
    Computers: [
        {
            title: 'Arch Music',
            link: '#',
            free: 16,
            total: 41,
        },
        {
            title: 'Biol Sci',
            link: '#',
            free: 60,
            total: 205,
        },
        {
            title: 'Central Library',
            link: '#',
            free: 13,
            total: 80,
        },
        {
            title: 'DHEngSci',
            link: '#',
            free: 41,
            total: 132,
        },
        {
            title: 'DuhigStudy',
            link: '#',
            free: 64,
            total: 196,
        },
        {
            title: 'Gatton',
            link: '#',
            free: 48,
            total: 101,
        },
        {
            title: 'Herston',
            link: '#',
            free: 22,
            total: 38,
        },
        {
            title: 'Law Library',
            link: '#',
            free: 15,
            total: 125,
        },
        {
            title: 'PACE',
            link: '#',
            free: 33,
            total: 58,
        },
        {
            title: 'Whitty',
            link: '#',
            free: 4,
            total: 9,
        },
    ],
    Training: [
        {
            date: 'Tuesday 22nd September 2020',
            day: 'Tuesday',
            dayDate: 22,
            monthDate: 'September',
            title: 'NVivo Pro - Next steps',
            time: '9:30am',
            format: 'Online',
            link: '#',
        },
        {
            date: 'Wednesday 23rd September 2020',
            day: 'Wednesday',
            dayDate: 23,
            monthDate: 'September',
            title: 'Excel: processing data',
            time: '9:30am',
            format: 'Online',
            link: '#',
        },
        {
            date: 'Wednesday 23rd September 2020',
            day: 'Wednesday',
            dayDate: 23,
            monthDate: 'September',
            title: 'Endnote: getting started',
            time: '1:30pm',
            format: 'Online',
            link: '#',
        },
        {
            date: 'Wednesday 23rd September 2020',
            day: 'Wednesday',
            dayDate: 23,
            monthDate: 'September',
            title: 'EndNote for thesis and publications writing',
            time: '1:30pm',
            format: 'Online',
            link: '#',
        },
        {
            date: 'Thursday 24th September 2020',
            day: 'Thursday',
            dayDate: 24,
            monthDate: 'September',
            title: 'R data manipulation with RStudio and dplyr:',
            time: '9:30am',
            format: 'Online',
            link: '#',
        },
        {
            date: 'Thursday 24th September 2020',
            day: 'Thursday',
            dayDate: 24,
            monthDate: 'September',
            title: 'Researcher Lunchbox Session: Whoâ€™s looking at your research?',
            time: '12:00pm',
            format: 'Online',
            link: '#',
        },
    ],
};
