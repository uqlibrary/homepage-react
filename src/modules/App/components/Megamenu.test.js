import { default as menuLocale } from 'locale/menu';

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

import { Megamenu } from './Megamenu';

const defaultMenuItems = [
    {
        linkTo: '/',
        primaryText: 'Primary text 0',
        secondaryText: 'secondary text 0',
    },
    {
        linkTo: '/xyz',
        primaryText: 'Primary text 1',
        secondaryText: 'secondary text 1',
    },
];

const defaultLocale = {
    skipNavTitle: 'Skip navigation',
    skipNavAriaLabel: 'Skip navigation',
    closeMenuLabel: 'Close menu',
};

function setup(testProps = {}) {
    const props = {
        ...testProps,
        classes: {},
        logoText: 'test',
        logoImage: 'test',
        logoLink: 'test',
        menuItems: testProps.menuItems || defaultMenuItems,
        toggleMenu: testProps.toggleMenu || jest.fn(),
        history: testProps.history || { push: jest.fn() },
        locale: testProps.locale || defaultLocale,
        menuOpen: testProps.menuOpen || false,
        docked: testProps.docked || false,
    };
    return getElement(Megamenu, props);
}

describe('Component Megamenu', () => {
    it('should render empty drawer', () => {
        const wrapper = setup();
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('should render opened drawer with menus', () => {
        const wrapper = setup({ menuOpen: true });
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('should display mobile correctly', () => {
        const wrapper = setup({ menuOpen: true, isMobile: true });
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    // it('should have working method for navigating to links', () => {
    // });
    //
    // it('should have working method for skipping menu items', () => {
    // });
    //
    // it('should have working method for focusing on given element ID', () => {
    // });

    it('should have a valid locale', () => {
        // because the menu locale is meant to be maintained by the user, we check the file is valid
        menuLocale.publicmenu.forEach(menu => {
            expect(menu.linkTo.length).not.toBe(0);
            expect(menu.primaryText.length).not.toBe(0);
            expect(menu.dataTestid.length).not.toBe(0);
            !menu.shiftLeft || expect(menu.shiftLeft).toEqual(true);
            !menu.submenuItems ||
                menu.submenuItems.forEach(submenu => {
                    expect(submenu.linkTo.length).not.toBe(0);
                    expect(submenu.primaryText.length).not.toBe(0);
                });
        });
    });
});
