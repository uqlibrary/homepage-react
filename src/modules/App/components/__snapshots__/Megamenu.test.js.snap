// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Megamenu should display mobile correctly 1`] = `
<div
  data-testid="mega-menu"
  id="megamenu"
>
  <WithStyles(ForwardRef(List))
    component="nav"
    data-testid="main-menu"
    id="mainMenu"
  >
    <div
      key="menucontainer-item-0"
    >
      <WithStyles(ForwardRef(ListItem))
        button={true}
        className="submenuheader"
        data-testid="megamenu-submenus-item-0"
        id="megamenu-submenus-item-0"
        key="megamenu-submenus-item-0"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemText))
          classes={
            Object {
              "primary": undefined,
              "secondary": undefined,
            }
          }
          primary="Primary text 0"
          secondary="secondary text 0"
        />
      </WithStyles(ForwardRef(ListItem))>
    </div>
    <div
      key="menucontainer-item-1"
    >
      <WithStyles(ForwardRef(ListItem))
        button={true}
        className="submenuheader"
        data-testid="megamenu-submenus-item-1"
        id="megamenu-submenus-item-1"
        key="megamenu-submenus-item-1"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemText))
          classes={
            Object {
              "primary": undefined,
              "secondary": undefined,
            }
          }
          primary="Primary text 1"
          secondary="secondary text 1"
        />
      </WithStyles(ForwardRef(ListItem))>
    </div>
  </WithStyles(ForwardRef(List))>
</div>
`;

exports[`Component Megamenu should render empty drawer 1`] = `
<div
  className="megamenu empty"
  data-testid="mega-menu-empty"
/>
`;

exports[`Component Megamenu should render opened drawer with menus 1`] = `
<div
  data-testid="mega-menu"
  id="megamenu"
>
  <WithStyles(ForwardRef(List))
    component="nav"
    data-testid="main-menu"
    id="mainMenu"
  >
    <div
      key="menucontainer-item-0"
    >
      <WithStyles(ForwardRef(ListItem))
        button={true}
        className="submenuheader"
        data-testid="megamenu-submenus-item-0"
        id="megamenu-submenus-item-0"
        key="megamenu-submenus-item-0"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemText))
          classes={
            Object {
              "primary": undefined,
              "secondary": undefined,
            }
          }
          primary="Primary text 0"
          secondary="secondary text 0"
        />
      </WithStyles(ForwardRef(ListItem))>
    </div>
    <div
      key="menucontainer-item-1"
    >
      <WithStyles(ForwardRef(ListItem))
        button={true}
        className="submenuheader"
        data-testid="megamenu-submenus-item-1"
        id="megamenu-submenus-item-1"
        key="megamenu-submenus-item-1"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(ListItemText))
          classes={
            Object {
              "primary": undefined,
              "secondary": undefined,
            }
          }
          primary="Primary text 1"
          secondary="secondary text 1"
        />
      </WithStyles(ForwardRef(ListItem))>
    </div>
  </WithStyles(ForwardRef(List))>
</div>
`;
